feat: Sentry support to automatically catch and report errors (opt-in)

akhilnarang opened the pull request:
- feat: Error reporting integrations
- build(deps): add sentry/browser dependency
- feat: client side error reporting + sentry
- feat: setup sentry integration

This requires setting of the environment variable `FRAPPE_SENTRY_DSN` as well as an explicit opt-in from the user's side via `System Settings` -> `Telemetry` or during setup wizard. 


`no-docs` Not really meant to be used by others, but feel free to use if you want to. Just setting DSN is all that's required.

ankush committed:
build(deps): add sentry/browser dependency

ankush committed:
feat: client side error reporting + sentry

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: setup sentry integration

Inspired primarily from sentry's generic WSGI integration

Environment variable `FRAPPE_SENTRY_DSN` needs to be enabled as well
as explicit opt-in from the user's side in system sid telemetry
settings

Conditionally include telemetry JS bundles

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/core/doctype/system_settings/system_settings.json:
```diff
@@ -634,12 +636,24 @@
+  },
+  {
+   "default": "0",
+   "fieldname": "auto_report_errors",
```
Lets reuse `enable_telemtry`

ankush commented on frappe/www/app.py:
```diff
@@ -46,6 +46,9 @@ def get_context(context):
 	include_css = hooks.get("app_include_css", []) + frappe.conf.get("app_include_css", [])
 	include_icons = hooks.get("app_include_icons", [])
 
+	if frappe.get_system_settings("auto_report_errors"):
+		include_js = hooks["error_reporting_js"] + include_js
```
lets avoid hook, and directly include telemetry/sentry bundle.

ankush commented on frappe/utils/sentry.py:
```diff
@@ -0,0 +1,77 @@
+		sentry_info := (frappe.conf.get("error_reporting", {})).get("sentry")
+	):
+		try:
+			init(sentry_info["dsn"], before_send=before_send)
```
Lets move this env variables, so we can init it only once.

ankush commented on frappe/utils/sentry.py:
```diff
@@ -0,0 +1,77 @@
+		return
+
+	if sentry_info := (frappe.conf.get("error_reporting", {})).get("sentry"):
+		bootinfo.sentry = sentry_info
```
reuse telemetry namepsace if possible

ankush commented on frappe/public/js/sentry.bundle.js:
```diff
@@ -0,0 +1,6 @@
+
+Sentry.init({
+	dsn: frappe.boot.sentry?.dsn,
+	release: frappe?.boot?.versions?.frappe,
```
Make sure release tracking works on both server/client side.

ankush commented on frappe/app.py:
```diff
@@ -395,6 +396,9 @@ def handle_exception(e):
 		print(frappe.get_traceback())
 
+	# Upload exception to Sentry
+	frappe.utils.sentry.capture_exception(e)
```
For now lets do this for `frappe.log_error`. Tracking all exception is a lot of notice.

ankush commented on frappe/app.py:
```diff
@@ -454,6 +454,14 @@ def serve(
 		log.setLevel(logging.ERROR)
 
+	# Always initialize sentry SDK if the DSN is sent
+	if sentry_dsn := os.getenv("SENTRY_DSN"):
```
This only runs for development server. We run using gunicorn in prod. Move this code at top level outside of any functions.

ankush replied:
Also prefer prefixing with "FRAPPE_" to avoid conflicting env variables

ankush commented on frappe/hooks.py:
```diff
@@ -31,6 +31,13 @@
+
+# JS code to include for error reporting
+# This is only loaded if error reporting is enabled.
+error_reporting_js = [
```
We can call this hook telemetry_js? Then it's useful for other apps too to conditionally load things.

ankush commented on frappe/utils/error.py:
```diff
@@ -70,6 +71,9 @@ def log_error(
 	capture("error_logged", "frappe", properties={"title": title, "trace_id": trace_id})
 
+	# Capture exception data if telemetry is enabled
+	capture_exception(message=f"{title}\n{traceback}")
```
This will trigger import of sentry even if it's not configured. Can avoid that memory cost by only importing it if sentry is configured.

Also if import is required, it's best to preload all those modules in app.py.

Refer https://github.com/frappe/frappe/pull/21475

akhilnarang committed:
fix(js): use the correct field for sentry DSN

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

Alchez commented on frappe/__init__.py:
```diff
@@ -59,6 +59,14 @@
+
+	from frappe.utils.sentry import before_send
+
+	sentry_sdk.init(dsn=sentry_dsn, before_send=before_send, release=__version__)
```
Should we also setup [tracing](https://docs.sentry.io/platforms/python/performance) here and allow that to be configured? Or is that planned for a future iteration?

We've gained some insights into backend performance from it (one of the projects below):

![image](https://github.com/frappe/frappe/assets/13396535/f986448a-744c-492e-be75-d56761931511)

ankush replied:
We have enabled tracing internally but we'll likely add it as an optional feature in future iterations. First cut is just for error reports.

Alchez commented on package.json:
```diff
@@ -25,6 +25,7 @@
     "@headlessui/vue": "^1.7.16",
     "@popperjs/core": "^2.11.2",
     "@redis/client": "^1.5.8",
+    "@sentry/browser": "^7.15.0",
```
Just curious since this version is over a year old: is this being pinned for compatibility?

ankush replied:
Hmm... @akhilnarang we should pick highest current supported version. Not sure if our deployment is compatible though.

ankush replied:
Anyway `^` means it will install latest minor?

So it seems correct only: https://www.npmjs.com/package/@sentry/browser 

See lock file. 

![image](https://github.com/frappe/frappe/assets/9079960/9be984c9-ee0a-4b7d-870f-016b4f2b176e)

akhilnarang replied:
> Not sure if our deployment is compatible though.

Most things should be fine since even latest python SDK is fine, besides the extra field being sent that's getting discarded for now.

Will bump this before merge, for now its older since I had just rebased the old branch.

ankush committed:
refactor: code cleanup

- better boot config name
- send sentry after - because frappe namespce doesn't exist if it starts
  first
- remove import in app.py because __init__ is always imported so no
  need.
- leave telemetry JS always present, this is used even when telemetry is
  not enabled.

ankush committed:
ci: report all errors to sentry

This can be useful.

ankush committed:
fix: tweak sentry config

- disable session tracking - This causes sentry to phone home on every load, we don't need this (yet)
- disable automatic integrtaions (like redis)
- disable all default integration and cherry-pick what's required

ankush commented:
note: port together - https://github.com/frappe/frappe/pull/23534
